README.TXT
cclpinfo (Chip's Challenge level pack info)     version 1.4

cclpinfo reads a Chip's Challenge file and displays the level names and
numbers. It can optionally display the time limit, password, chips required,
and level hints as well.

Usage:  cclpinfo [-ptch] files...

    files   one or more levelsets to read
    -p      show passwords
    -t      show time limits
    -c      show chips required/available
    -h      show hints

When the -c option is used, if the number of required chips is the
same as the number of available chips, only one number is displayed;
but if they are different, they are both displayed.

For example:

    103. Island Hopping                             30
    104. Pyramid                                    66/84

Island Hopping from the original levelset requires 30 chips and has 30 chips
available. Pyramid has 84 chips available but only 66 are required to open the
chip socket.

If the file is not a valid Chip's Challenge file, cclpinfo will abort. 
Whether or not the file is a valid Chip's Challenge file is determined 
by looking for a "signature" at the beginning of the file. The name of 
the file does not matter.

Tip: If you want to save the output to a file, add " > filename" to the end
of the line. Be careful: if the file already exists, the contents will be 
erased. To append to an existing file, add " >> filename" instead.

Example:

    cclpinfo CHIPS.dat -pt > chipsinfo.txt

This will make (or overwrite) a file called "chipsinfo.txt" in the current 
directory, and into the file will be written a list of numbers, titles, 
passwords, and time limits for every level in "CHIPS.dat"

-----------------------------------------------------------------------

cclpinfo was written by Andrew E. It is based off a program called datstat
which was written by Madhav Shanbhag and posted to the Chip's Challenge
newsgroup.

You can find the latest version of cclpinfo at 
https://github.com/magical/cclpinfo/
or
https://tilde.town/~magical/chip/#cclpinfo

Please report bugs on github
https://github.com/magical/cclpinfo/issues/new

The windows executable was compiled with MinGW gcc v6.3.0 on Fedora 25.


CHANGELOG.TXT
version 1.5     2017-08-??
--------------------------

New features:

Bug fixes:
- fixed a memory leak when a level had multiple title, password, or hint fields

version 1.4	2017-08-27
-----------------------------

New features:
- cclpinfo has a new home on github: https://github.com/magical/cclpinfo
- you can now give cclpinfo multiple files on the command line and it will list
  all of them
- cclpinfo can now read levelsets that use pgchip

Bug fixes:
- rewrote most of the code
- fixed signature checking on 64-bit architectures
- removed some assumptions about endianness -- cclpinfo should now work on
  big-endian architectures
- fixed heap corruption when parsing certain invalid files (found with afl-fuzz)

version 1.3	2006-01-09
-----------------------------
The -c option now not only displays the chips required, but also the
number of chips in the level. If the number of required chips is the
same as the number of available chips, only one number is displayed;
but if they are different, they are both displayed. Example:

103. Island Hopping                             30
104. Pyramid                                    66/84


version 1.2	2006-01-04
-----------------------------
Added info about how to save output to a file to the readme.
Added info about how cclpinfo determines if a file is valid.
cclpinfo now displays the filename, for convenience when the output is 
saved to a file.
New feature: 
 -c option: display chips required to open the chip socket


version 1.1	2006-01-03
-----------------------------
Changed program slightly to display pi^2.dat correctly.
Changed program name to "cclpinfo".


version 1.0	2006-01-02
-----------------------------
Initial release under name "newstat".
New features:
 - password display
 - time limit display
 - hint display
 - checks signature of file


datstat version 1.0	2006-01-01
-----------------------------
Written by Madhav Shanbhag, who posted it on the Chip's Challenge newsgroup.


datstat.c
#include <stdio.h>
#include <stdlib.h>


typedef unsigned char byte;
typedef unsigned short int word;
typedef unsigned long int dword;


struct LevelInfo
{
	word number;
	word time;
	word chips;
} levelInfo;


void displayLevelStats(const byte* pLevelData, word levelSize)
	{
	const byte* p = pLevelData;
	int i;
	#define bump(T) *(((const T*)p)++)

	levelInfo = bump(struct LevelInfo);
	p += sizeof(word);
	for (i = 0; i < 2; ++i)
		{
		word layerSize = bump(word);
		p += layerSize;
		}
	p += sizeof(word) + sizeof(byte) + sizeof(byte);
	printf("%3d. %s\n", levelInfo.number, p);

	#undef bump
	}


int processFile(const char* szDatFileName)
	{
	FILE* fp = NULL;
	word nLevels = 0, l;

	fp = fopen(szDatFileName, "rb");
	if (!fp)
		{
		printf("Unable to open file: %s\n", szDatFileName);
		return 1;
		}

	fseek(fp, sizeof(dword), SEEK_SET);
	fread(&nLevels, sizeof(word), 1, fp);

	for (l = 1; l <= nLevels; ++l)
		{
		word levelSize = 0;
		byte* pLevelData = NULL;

		fread(&levelSize, sizeof(word), 1, fp);
		if (levelSize == 0) break;

		pLevelData = malloc(levelSize);
		fread(pLevelData, levelSize, 1, fp);
		displayLevelStats(pLevelData, levelSize);
		free(pLevelData);
		}

	fclose(fp);
	return 0;
	}


int main(int nParams, const char* szParam[])
	{
	if (nParams != 1+1)
		{
		puts("Syntax:   datstat  datfile [> outfile]");
		return -1;
		}

	return processFile(szParam[1]);
	}


newstat.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define PASSWORD_MASK 0x99

typedef unsigned char byte;
typedef unsigned short int word;
typedef unsigned long int dword;


struct
{
	word number;
	word time;
	word chips;
	char *title;
	char *password;
	char *hint;
} levelInfo;

struct
{
	const char *datfile_name;
	bool display_passwords:1;
	bool display_time:1;
	bool display_hints:1;
} options;

void displayLevelStats(FILE* fp)
	{
	int i = 0;
	char *j;
	word length;

	levelInfo.title = NULL;
	levelInfo.password = NULL;
	levelInfo.hint = NULL;

	fread(&levelInfo.number, sizeof(word), 1, fp); // the level number
	fread(&levelInfo.time, sizeof(word), 1, fp); // the time limit
	fread(&levelInfo.chips, sizeof(word), 1, fp); // # chips required
	fseek(fp, sizeof(word), SEEK_CUR); // 0x0001
	fread(&length, sizeof(word), 1, fp); // length of upper layer data
	fseek(fp, length, SEEK_CUR); // we don't care about the level data
	fread(&length, sizeof(word), 1, fp); // length of lower layer data
	fseek(fp, length, SEEK_CUR);
	fread(&length, sizeof(word), 1, fp); // length of misc data

	i = 0;
	while (i < length)
		{
		byte fieldnum, fieldlength;
		fread(&fieldnum, sizeof(byte), 1, fp);
		fread(&fieldlength, sizeof(byte), 1, fp);
		switch (fieldnum)
			{
			// 1 == time limit
			// 2 == chips
			case 3:
				levelInfo.title = malloc(fieldlength);
				fread(levelInfo.title, fieldlength, 1, fp);
				break;
			// 4 == trap connections
			// 5 == clone connections
			case 6:
				levelInfo.password = malloc(fieldlength);
				fread(levelInfo.password, fieldlength, 1, fp);
				j = levelInfo.password;
				while (*j)
					{
					*j = *j ^ PASSWORD_MASK;
					j++;
					}
				//printf("%s", levelInfo.password);
				break;
			case 7:
				levelInfo.hint = malloc(fieldlength);
				fread(levelInfo.hint, fieldlength, 1, fp);
				break;
			// 8 and 9 are purposeless
			// 10 == creature list
			default:
				fseek(fp, fieldlength, SEEK_CUR);
			}
		i += fieldlength + 2;
		}
	
	printf("%3d. %-50s", levelInfo.number, levelInfo.title);
	if (options.display_passwords)
		{
		printf("\t%s", levelInfo.password);
		}
	if (options.display_time)
		{
		if (levelInfo.time == 0)
			{
			printf("\t---");
			}
		else
			{
			printf("\t%03i", levelInfo.time);
			}
		}
	fputc('\n', stdout);

	if (options.display_hints && levelInfo.hint != NULL)
		{
		printf("     %s\n", levelInfo.hint);
		}

	if (levelInfo.title != NULL)
		{
		free(levelInfo.title);
		}
	if (levelInfo.password != NULL)
		{
		free(levelInfo.password);
		}
	if (levelInfo.hint != NULL)
		{
		free(levelInfo.hint);
		}
	}


int processFile(const char* szDatFileName)
	{
	FILE* fp = NULL;
	word nLevels = 0, l;
	dword signature;

	fp = fopen(szDatFileName, "rb");
	if (!fp)
		{
		printf("Unable to open file: %s\n", szDatFileName);
		return 1;
		}
	
	fread(&signature, sizeof(dword), 1, fp);
	if (signature != 0x0002aaac && signature != 0x0102aaac)
		{
		fprintf(stderr, "%s is not a valid Chip's Challenge file.\n", szDatFileName);
		fclose(fp);
		return 1;
		}
	fread(&nLevels, sizeof(word), 1, fp);
	
	printf("  #. %-50s%s%s\n", "Title", (options.display_passwords?"\tPass":""), (options.display_time?"\tTime":""));
	for (l = 1; l <= nLevels; ++l)
		{
		word levelSize = 0;
		fread(&levelSize, sizeof(word), 1, fp);
		if (levelSize == 0) break;
		displayLevelStats(fp);
		}

	fclose(fp);
	return 0;
	}

int main(int argc, const char *argv[])
	{
	int i, j;
	
	if (argc <= 1)
		{
		puts("Usage:\tdatstat datfile [-pth]");
		puts("");
		puts("\tdatfile\tThe location of the .dat file to use");
		puts("\tp\tDisplay passwords");
		puts("\tt\tDisplay the time limits");
		puts("\th\tDisplay the level hints");
		return -1;
		}
	
	options.datfile_name = NULL;
	options.display_passwords = false;
	options.display_time = false;
	options.display_hints = false;
	//parse the command-line options
	for (i = 1; i < argc; i+=1)
		{
		if (argv[i][0] == '-')
			{
			for (j = 1; j < strlen(argv[i]); j+=1)
				{
				switch (argv[i][j])
					{
					case 'p':
						if (options.display_passwords == true)
							{
							fputs("Warning: option -p is used more than once\n", stderr);
							}
						options.display_passwords = true;
						break;
					case 't':
						if (options.display_time == true)
							{
							fputs("Warning: option -t is used more than once\n", stderr);
							}
						options.display_time = true;
						break;
					case 'h':
						if (options.display_hints == true)
							{
							fputs("Warning: option -h is used more than once\n", stderr);
							}
						options.display_hints = true;
						break;
					default:
						fprintf(stderr, "Warning: option -%c is invalid\n", argv[i][j]);
					}
				}
			}
		else if (options.datfile_name == NULL)
			{
			options.datfile_name = argv[i];
			}
		else
			{
			}
		}
	if (options.datfile_name == NULL)
		{
		fputs("Error: no datfile is specified", stderr);
		return -1;
		}

	return processFile(options.datfile_name);
	}


cclpinfo.c
/* cclpinfo v1.4 */

#include <stdarg.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define PASSWORD_MASK 0x99
#define CHIP_TILE 0x02
#define LAYER_SIZE (32 * 32)

struct file
{
	FILE *fp;
	const char *name;
	int pos;
};

struct levelinfo
{
	int number;
	int time;
	int chips;
	int totalchips;
	char *title;
	char *password;
	char *hint;
	int upperlayersize;
	int lowerlayersize;
};

struct
{
	bool showpassword;
	bool showtime;
	bool showhint;
	bool showchips;
} options;

uint8_t readbyte(struct file *file)
{
	uint8_t buf[1] = {};
	if (fread(buf, sizeof buf, 1, file->fp) != 1) {
		return 0;
	}
	file->pos += 1;
	return buf[0];
}

uint16_t readword(struct file *file)
{
	uint8_t buf[2] = {};
	if (fread(buf, sizeof buf, 1, file->fp) != 1) {
		return 0;
	}
	file->pos += 2;
	return ((uint16_t)buf[0]) | ((uint16_t)buf[1] << 8);
}

void readstring(struct file *file, int len, char **bufp, char *debug_name)
{
	char *buf;
	ssize_t n;

	buf = *bufp = realloc(*bufp, len + 1);
	n = fread(buf, 1, len, file->fp);
	file->pos += n;
	if (n > 0 && buf[n - 1] != '\0') {
		fprintf(stderr, "%s:%d: warning: %s not null terminated\n", file->name, file->pos, debug_name);
	}
	buf[n] = '\0';
}

void skipbytes(struct file *file, int n)
{
	if (fseek(file->fp, n, SEEK_CUR) == 0) {
		file->pos += n;
	} else {
		//perror("fseek");
		// unseekable file
		for (; n > 0; n--) {
			if (fgetc(file->fp) == EOF) {
				break;
			}
			file->pos += 1;
		}
	}
}

void warnf(const char *fmt, ...)
{
	va_list va;
	va_start(va, fmt);
	vfprintf(stderr, fmt, va);
	va_end(va);
	//abort();
}

int count_tiles(struct file *file, int layersize, int search_tile, char *layer_name)
{
	int total = 0;
	int count = 0;
	int c, tile;

	int start = file->pos;

	while (layersize > 0) {
		c = fgetc(file->fp);
		if (c == EOF) {
			warnf("%s:%d: warning: unexpected end of file\n", file->name, file->pos);
			break;
		}
		file->pos += 1;
		if (c == 0xff) {
			c = readbyte(file);
			tile = readbyte(file);
			if (tile == search_tile) {
				count += c;
			}
			total += c;
			layersize -= 3;
		} else {
			if (c == search_tile) {
				count += 1;
			}
			total += 1;
			layersize -= 1;
		}
	}

	if (total != LAYER_SIZE) {
		warnf("%s:%d: warning: %s layer has %d tiles, expected %d\n", file->name, start, layer_name, total, LAYER_SIZE);
	}

	return count;
}

void readlevel(struct file *file, off_t levelsize, struct levelinfo *info)
{
	int l = levelsize;
	int miscsize;
	int i;

	info->number = readword(file); // the level number
	info->time = readword(file);   // the time limit
	info->chips = readword(file);  // # chips required
	readword(file);                // 0x0001
	l -= 2 * 4;

	info->totalchips = 0;
	info->upperlayersize = readword(file); // length of upper layer data
	info->totalchips += count_tiles(file, info->upperlayersize, CHIP_TILE, "upper");
	info->lowerlayersize = readword(file); // length of lower layer data
	info->totalchips += count_tiles(file, info->lowerlayersize, CHIP_TILE, "lower");

	l -= 2 + info->upperlayersize;
	l -= 2 + info->lowerlayersize;

	miscsize = readword(file); // length of misc data
	l -= 2;

	if (miscsize != l) {
		warnf("%s: mismatch between level size and misc data size\n", file->name);
	}

	if (miscsize > 1152) {
		warnf("%s: misc data size exceeds 1152 bytes; MSCC will not be able to load this level\n", file->name);
	}

	while (l > 0) {
		int fieldnum, fieldlength;
		fieldnum = readbyte(file);
		fieldlength = readbyte(file);
		switch (fieldnum) {
		// 1: time limit
		// 2: chips
		case 3: // level title
			readstring(file, fieldlength, &info->title, "level title");
			break;
		// 4: trap connections
		// 5: clone connections
		case 6: // password
			readstring(file, fieldlength, &info->password, "password");
			for (i = 0; i < fieldlength && info->password[i] != '\0'; i++) {
				info->password[i] ^= PASSWORD_MASK;
			}
			break;
		case 7: // level hint
			readstring(file, fieldlength, &info->hint, "hint");
			break;
		// I don't know what 8 and 9 are for
		// 10: creature list
		default:
			skipbytes(file, fieldlength);
		}
		l -= fieldlength + 2;
	}
}

void freelevel(struct levelinfo *info)
{
	free(info->title);
	free(info->password);
	free(info->hint);

	info->title = NULL;
	info->password = NULL;
	info->hint = NULL;
}

void printlevel(FILE *out, struct levelinfo *info)
{
	fprintf(out, "%3d. %-35s ", info->number, info->title);
	if (options.showpassword) {
		fprintf(out, "\t%s", info->password);
	}
	if (options.showtime) {
		if (info->time == 0) {
			fprintf(out, "\t---");
		} else {
			fprintf(out, "\t%03d", info->time);
		}
	}
	if (options.showchips) {
		if (info->chips == info->totalchips) {
			fprintf(out, "\t%d", info->chips);
		} else {
			fprintf(out, "\t%d/%d", info->chips, info->totalchips);
		}
	}

	fprintf(out, "\n");

	if (options.showhint && info->hint != NULL) {
		fprintf(out, "     %s\n", info->hint);
	}
}

const char *extract_filename(const char *path)
{
	const char *filename = path;
	for (; *path; path++) {
		if (*path == '/' || *path == '\\') {
			filename = path + 1;
		}
	}
	return filename;
}

int processFile(const char *filename)
{
	FILE *fp = NULL;
	struct file file = {};
	struct levelinfo info = {};
	int nLevels, i;
	uint16_t signature, version;

	fp = fopen(filename, "rb");
	if (!fp) {
		printf("Unable to open file: %s\n", filename);
		return 1;
	}

	file.fp = fp;
	file.name = filename;
	file.pos = 0;

	signature = readword(&file);
	version = readword(&file);
	if (signature != 0xaaac || (version != 0x0002 && version != 0x0102 && version != 0x0003)) {
		fprintf(stderr, "%s is not a valid Chip's Challenge file.\n", filename);
		fclose(fp);
		return 1;
	}

	nLevels = readword(&file);

	printf("%s\n\n", extract_filename(filename));

	printf("  #. %-35s%s%s%s\n", "Title",
	       (options.showpassword ? "\tPass" : ""),
	       (options.showtime ? "\tTime" : ""),
	       (options.showchips ? "\tChips" : ""));

	for (i = 1; i <= nLevels; i++) {
		int levelsize = readword(&file);
		if (levelsize == 0)
			break;
		readlevel(&file, levelsize, &info);
		printlevel(stdout, &info);
		freelevel(&info);
	}

	printf("\n");

	fclose(fp);
	return 0;
}

void usage(const char *program)
{
	puts("cclpinfo v1.4 written by Andrew E. and Madhav Shanbhag");
	puts("");
	printf("Usage:\t%s [-ptch] files...\n", program);
	puts("");
	puts("\tfiles\tone or more levelsets to read");
	puts("\t-p\tshow passwords");
	puts("\t-t\tshow time limits");
	puts("\t-c\tshow chips required/available");
	puts("\t-h\tshow hints");
	puts("");
	puts("For more information, read README.TXT");
}

int main(int argc, const char *argv[])
{
	int i, j;
	int nfiles;
	int err;

	if (argc <= 1) {
		usage(argv[0]);
		return 2;
	}

	// initialize the options
	options.showpassword = false;
	options.showtime = false;
	options.showchips = false;
	options.showhint = false;

	//parse the command-line options
	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			for (j = 1; argv[i][j] != '\0'; j++) {
				switch (argv[i][j]) {
				case 'p':
					options.showpassword = true;
					break;
				case 't':
					options.showtime = true;
					break;
				case 'c':
					options.showchips = true;
					break;
				case 'h':
					options.showhint = true;
					break;
				default:
					fprintf(stderr, "Warning: unknown option -%c\n", argv[i][j]);
				}
			}
		}
	}

	err = 0;
	nfiles = 0;
	for (i = 1; i < argc; i++) {
		if (argv[i][0] != '-') {
			nfiles++;
			if (processFile(argv[i]) != 0) {
				err = 1;
			}
		}
	}

	if (nfiles == 0) {
		fprintf(stderr, "error: no files specified");
		return 2;
	}

	return err;
}
